core/src/main/scala/fs2/concurrent.scala

```diff
@@ -23,8 +44,10 @@ object concurrent {
               }
               else F.pure(())
             }}}
-          s.chunks.attempt.evalMap { o => outputQueue.enqueue1(Some(o)) }
-           .interruptWhen(killSignal).run.flatMap { _ => Free.eval(done) }.run
+          inner.chunks.attempt.evalMap { o => outputQueue.enqueue1(Some(o)) }
+          .interruptWhen(killSignal)
+          .onFinalize(done)
+          .run.run
         }}
```

Comment:
yes 100% positive it does.  I was  starring on it same way, and took me half day to pint point it. It may uncover issue somewhere else. But it really does. I will be putting out the tcp PR laters today, so you may try to run it w/o this change.